File: app/src/main/java/com/coinninja/coinkeeper/db/Migration.java
package com.coinninja.coinkeeper.db;

import org.greenrobot.greendao.database.Database;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public interface Migration {
    /**
     * Applies migration to provided database
     *
     * @param db
     * @param currentVersion version prior to migration
     * @return version after migration has been applied
     */
    int runMigration(@NonNull Database db, int currentVersion);

    /**
     * @return null or an instance of current migration
     */
    @Nullable
    Migration getPreviousMigration();

    int getTargetVersion();

    /**
     * @return the new version that will result
     */
    int getMigratedVersion();
}

File: app/src/main/java/com/coinninja/coinkeeper/db/Migration.kt
package com.coinninja.coinkeeper.db

import org.greenrobot.greendao.database.Database

interface Migration {
    /**
     * Applies migration to provided database
     *
     * @param db
     * @param currentVersion version prior to migration
     * @return version after migration has been applied
     */
    fun runMigration(db: Database, currentVersion: Int): Int

    /**
     * @return null or an instance of current migration
     */
    val previousMigration: Migration?

    val targetVersion: Int
    /**
     * @return the new version that will result
     */
    val migratedVersion: Int
}